engine/execution/storehouse/checkpoint_validator.go

```diff
@@ -0,0 +1,137 @@
+package storehouse
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-go/engine/execution"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/convert"
+	"github.com/onflow/flow-go/ledger/complete/wal"
+	"github.com/onflow/flow-go/storage"
+)
+
+// ValidateWithCheckpoint validates the registers in the given store against the leaf nodes read from the checkpoint file.
+// Limitation: the validation can not cover if there are extra non-empty registers in the store that are not in the checkpoint file.
+func ValidateWithCheckpoint(
+	log zerolog.Logger,
+	ctx context.Context,
+	store execution.OnDiskRegisterStore,
+	results storage.ExecutionResults,
+	headers storage.Headers,
+	checkpointDir string, // checkpointDir must have a root.checkpoint file that contains only a single trie
+	blockHeight uint64,
+	workerCount int,
+) error {
+	// used by the wal reader to send leaf nodes read from checkpoint file
+	// used by N workers to validate registers in store
+	leafNodeChan := make(chan *wal.LeafNode, 1000)
+
+	// create N workers to validate registers in store
+	cct, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	g, gCtx := errgroup.WithContext(cct)
+
+	start := time.Now()
+	log.Info().Msgf("validation registers from checkpoint with %v worker", workerCount)
+	for i := 0; i < workerCount; i++ {
+		g.Go(func() error {
+			return validatingRegisterInStore(gCtx, store, leafNodeChan, blockHeight)
+		})
```

Comment:
Since `validatingRegisterInStore` returns on first error, do we want to cancel `gCtx` so other goroutines can exit earlier?

Reply:
Thanks for resolving the deadlock with commit [9871585](https://github.com/onflow/flow-go/pull/8257/commits/98715851936bcc7c65e9a142c964e7c6b93846d6) :+1: